File: .oldC++/kernel/src/knowledge/graph_builder.cpp
#include "jarvis/knowledge/graph_builder.h"
#include <QSqlDatabase>
#include <QSqlQuery>
#include <QSqlError>
#include <QJsonDocument>
#include <QJsonObject>
#include <QJsonArray>
#include <QString>
#include <set>

namespace jarvis::knowledge {

class GraphBuilder : public IGraphBuilder {
public:
    explicit GraphBuilder(const std::string& dbPath)
        : dbPath_(dbPath)
    {
        initDb();
    }

    GraphData build() override {
        GraphData graph;
        std::set<std::string> addedNodes;

        QSqlQuery nq(QSqlDatabase::database("knowledge-graph"));
        nq.exec("SELECT id, title, folder, tags FROM notes ORDER BY title");
        while (nq.next()) {
            GraphNode node;
            node.id = nq.value(0).toString().toStdString();
            node.label = nq.value(1).toString().toStdString();
            node.folder = nq.value(2).toString().toStdString();

            auto tagsStr = nq.value(3).toString().toStdString();
            if (!tagsStr.empty()) {
                QStringList parts = QString::fromStdString(tagsStr).split(',', Qt::SkipEmptyParts);
                for (const auto& p : parts)
                    node.tags.push_back(p.trimmed().toStdString());
            }

            // Count links for this node
            QSqlQuery lcq(QSqlDatabase::database("knowledge-graph"));
            lcq.prepare("SELECT COUNT(*) FROM note_links WHERE source_id = ? OR target_id = ?");
            lcq.addBindValue(QString::fromStdString(node.id));
            lcq.addBindValue(QString::fromStdString(node.id));
            if (lcq.exec() && lcq.next())
                node.linkCount = lcq.value(0).toInt();

            graph.nodes.push_back(std::move(node));
            addedNodes.insert(graph.nodes.back().id);
        }

        QSqlQuery eq(QSqlDatabase::database("knowledge-graph"));
        eq.exec("SELECT source_id, target_id, link_type FROM note_links");
        while (eq.next()) {
            GraphEdge edge;
            edge.source = eq.value(0).toString().toStdString();
            edge.target = eq.value(1).toString().toStdString();
            edge.linkType = eq.value(2).toString().toStdString();

            // Only include edges where both nodes exist
            if (addedNodes.count(edge.source) && addedNodes.count(edge.target))
                graph.edges.push_back(std::move(edge));
        }

        return graph;
    }

    std::string buildJson() override {
        auto graph = build();

        QJsonObject root;
        QJsonArray nodesArr;
        for (const auto& n : graph.nodes) {
            QJsonObject nodeObj;
            nodeObj["id"] = QString::fromStdString(n.id);
            nodeObj["label"] = QString::fromStdString(n.label);
            nodeObj["folder"] = QString::fromStdString(n.folder);
            QJsonArray tagsArr;
            for (const auto& t : n.tags)
                tagsArr.append(QString::fromStdString(t));
            nodeObj["tags"] = tagsArr;
            nodeObj["linkCount"] = n.linkCount;
            nodesArr.append(nodeObj);
        }
        root["nodes"] = nodesArr;

        QJsonArray edgesArr;
        for (const auto& e : graph.edges) {
            QJsonObject edgeObj;
            edgeObj["source"] = QString::fromStdString(e.source);
            edgeObj["target"] = QString::fromStdString(e.target);
            edgeObj["linkType"] = QString::fromStdString(e.linkType);
            edgesArr.append(edgeObj);
        }
        root["edges"] = edgesArr;

        return QJsonDocument(root).toJson(QJsonDocument::Compact).toStdString();
    }

private:
    std::string dbPath_;

    void initDb() {
        QString connName = "knowledge-graph";
        QSqlDatabase db;
        if (QSqlDatabase::contains(connName)) {
            db = QSqlDatabase::database(connName);
        } else {
            db = QSqlDatabase::addDatabase("QSQLITE", connName);
        }
        db.setDatabaseName(QString::fromStdString(dbPath_));
        if (!db.isOpen()) db.open();
    }
};

IGraphBuilder* createGraphBuilder(const std::string& dbPath) {
    return new GraphBuilder(dbPath);
}

} // namespace jarvis::knowledge

File: backend/jarvis/graph_builder.py
"""Standalone knowledge graph builder — nodes, edges, JSON export."""

import json
from dataclasses import dataclass, field

from jarvis.database import Database


@dataclass
class GraphNode:
    id: str = ""
    label: str = ""
    folder: str = ""
    tags: list[str] = field(default_factory=list)
    link_count: int = 0


@dataclass
class GraphEdge:
    source: str = ""
    target: str = ""
    link_type: str = "wikilink"


@dataclass
class GraphData:
    nodes: list[GraphNode] = field(default_factory=list)
    edges: list[GraphEdge] = field(default_factory=list)


def _parse_tags(tags_str: str) -> list[str]:
    if not tags_str:
        return []
    return [t.strip() for t in tags_str.split(",") if t.strip()]


class GraphBuilder:
    def __init__(self, db: Database):
        self._db = db

    def build(self) -> GraphData:
        note_rows = self._db.fetchall(
            "SELECT id, title, folder, tags FROM notes ORDER BY title"
        )
        edge_rows = self._db.fetchall(
            "SELECT source_id, target_id, link_type FROM note_links"
        )

        link_count: dict[str, int] = {}
        for e in edge_rows:
            link_count[e["source_id"]] = link_count.get(e["source_id"], 0) + 1
            link_count[e["target_id"]] = link_count.get(e["target_id"], 0) + 1

        node_ids: set[str] = set()
        nodes: list[GraphNode] = []
        for r in note_rows:
            nid = r["id"]
            node_ids.add(nid)
            tags = _parse_tags(r["tags"])
            nodes.append(GraphNode(
                id=nid,
                label=r["title"],
                folder=r["folder"],
                tags=tags,
                link_count=link_count.get(nid, 0),
            ))

        edges: list[GraphEdge] = []
        for e in edge_rows:
            if e["source_id"] in node_ids and e["target_id"] in node_ids:
                edges.append(GraphEdge(
                    source=e["source_id"],
                    target=e["target_id"],
                    link_type=e["link_type"],
                ))

        return GraphData(nodes=nodes, edges=edges)

    def build_json(self) -> str:
        graph = self.build()

        nodes_list = []
        for n in graph.nodes:
            node_obj: dict = {
                "id": n.id,
                "label": n.label,
                "folder": n.folder,
                "tags": n.tags,
                "linkCount": n.link_count,
            }
            nodes_list.append(node_obj)

        edges_list = []
        for e in graph.edges:
            edge_obj: dict = {
                "source": e.source,
                "target": e.target,
                "linkType": e.link_type,
            }
            edges_list.append(edge_obj)

        return json.dumps(
            {"nodes": nodes_list, "edges": edges_list},
            ensure_ascii=False,
        )
